TrafficSimulator/Assets/Cam/DefaultCamera.cs

```diff
@@ -0,0 +1,151 @@
+            }
+        }
+
+        public override void HandleDoubleClickInput(InputAction.CallbackContext ctx)
+        {
+            var ray = Camera.main.ScreenPointToRay(Mouse.current.position.ReadValue());
+            if (Physics.Raycast(ray, out var hitInfo) && hitInfo.transform.gameObject.Equals(_toggledGameObject))
+            {
+                CameraManager.CameraTarget = FollowTransform;
+                CameraManager.ToggleThirdPersonCamera();
+            }
+        }
+
+        public override void Move(Vector3 direction)
+        {
+            // Ignore the input argument if the mouse is near the screen border
+            if (_isNearScreenBorder) direction = _mousePointDirection.normalized;
+
+            if (direction.sqrMagnitude != 0 && _hasToggledGameObject) SetToggledGameObjectToNull();
+            Vector3 translatedDirection = TranslateDirectionToForward(direction.y, direction.x);
+            FollowTransform.position += translatedDirection * (_movementSpeed * Time.deltaTime);
+        }
+
+        public override void RotateHorizontal(float horizontalRotation)
+        {
+            FollowTransform.eulerAngles += new Vector3(0, horizontalRotation * _rotationSpeed * Time.deltaTime, 0);
+        }
+
+        public override void Zoom(float zoomValue)
+        {
+            _targetZoom -= zoomValue;
+            _targetZoom = Mathf.Clamp(_targetZoom, _maxZoom, _minZoom);
+            VirtualCamera.m_Lens.FieldOfView = Mathf.Lerp(VirtualCamera.m_Lens.FieldOfView,
+                _targetZoom, Time.deltaTime * _zoomSpeed);
+        }
+        
+        private void SetToggledGameObjectToNull()
+        {
+            _toggledGameObject.GetComponent<Outline>().enabled = false;
+            _toggledGameObject = null;
+            _hasToggledGameObject = false;
+        }
+
+        private void SetToggledGameObject(GameObject toggledGameObject)
+        {
+            Outline outline = toggledGameObject.GetComponent<Outline>();
+            outline.enabled = true;
+            _toggledGameObject = toggledGameObject;
+            _hasToggledGameObject = true;
+            StartCoroutine(PanToTarget(toggledGameObject.transform.position));
+        }
+        
+        private Vector3 TranslateDirectionToForward(float forwardScalar, float sidewaysScalar)
+        {
+            return FollowTransform.forward * forwardScalar + transform.right * sidewaysScalar;
+        }
+
+        IEnumerator PanToTarget(Vector3 target)
+        {
+            _isMovingTowardsToggledObject = true;
+            float startTime = Time.time;
+            float journeyLength = Vector3.Distance(FollowTransform.position, _toggledGameObject.transform.position);
+
+            while (Vector3.Distance(FollowTransform.transform.position, target) > 0.01f && _hasToggledGameObject)
+            {
+                float distCovered = (Time.time - startTime) * _togglePanSpeed;
+                float fractionOfJourney = distCovered / journeyLength;
```

Comment:
Could this yield a DivisionByZeroException if you try to pan to a target when `FollowTransform.position == _toggledGameObject.transform.position`?

Reply:
No since its an Enumerator the journey length will be set one time at line 138:
`float journeyLength = Vector3.Distance(FollowTransform.position, _toggledTransform.position);`

and then this initial value will be checked at line 140:
`while (Vector3.Distance(FollowTransform.position, target) > 0.01f && _hasToggledTransform)`

If journeyLength is below 0.01 the while-loop will never execute. 

Worth noting is indeed the fact that the function is called as an IEnumerator, that is, it is a function that can pause its execution and return where it left of with local variables and state preserved. Will not be a performance issue since it is not a high frequent call.

Reply:
No since its an Enumerator the journey length will be set one time at line 138:
`float journeyLength = Vector3.Distance(FollowTransform.position, _toggledTransform.position);`

and then this initial value will be checked at line 140:
`while (Vector3.Distance(FollowTransform.position, target) > 0.01f && _hasToggledTransform)`

If journeyLength is below 0.01 the while-loop will never execute. This was indeed not obvious since I used to different variables for the same Vector3 reference (target and _toggledGameObject.transform.position). This has been fixed now and the function only uses target such that:

```
float journeyLength = Vector3.Distance(FollowTransform.position, target);

while (Vector3.Distance(FollowTransform.position, target) > 0.01f && _hasToggledTransform)
```

Worth noting is indeed the fact that the function is called as an IEnumerator, that is, it is a function that can pause its execution and return where it left of with local variables and state preserved. Will not be a performance issue since it is not a high frequent call.

Reply:
Good explanation, thanks!